### PR TITLE
Optimizations for main build workflow

### PR DESCRIPTION
**Pull request by hairyhum**
## Change Overview

Trying to reduce time it takes to run the CI workflow. 
Currently packing and unpacking entire directory takes a long time. It's faster to just download the go dependencies.

Current changes:
- Don't pack entire dir on gomod job, only pass the go.mod and go.sum files as outputs
- Make `lint`, `test`, `build` and `docs` steps run in parallel
- Remove double invocation of `go test` in `test.sh`, this was left over from the time when test build was separate from test run, no longer required

These changes do not include any caching proposed in https://github.com/kanisterio/kanister/pull/2352 but already cut the pipeline time more then 2 times so it make sense to merge that first and then work on caching as a follow-up

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [x] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [ ] :sunflower: Feature
- [ ] :world_map: Documentation
- [ ] :robot: Test

## Test Plan

Run the pipeline.

- [ ] :muscle: Manual
- [x] :zap: Unit test
- [ ] :green_heart: E2E

### CODE REVIEW

File: .github/workflows/main.yaml [hunk 1]
```diff
@@ -8,37 +9,52 @@ on:
...
+          echo FILE
+        } >> "$GITHUB_OUTPUT"
+    - id: gosum
+      run: |
+        {
+          echo 'gosum<<FILE'
+          cat go.sum
+          echo
+          echo FILE
+        } >> "$GITHUB_OUTPUT"
```
Comment (viveksinghggits) [inline .github/workflows/main.yaml:33]: earlier we used to upload the code from one step and download it on other step? How does that work now? I see we are just outputting the go.* files and getting them from other step. How is code available in other steps?

Also, it looks like we are not running `go mod download` anymore, do we not need that?

Last question :smile: , why do we have to just copy the go.* between steps and not other code.

Comment (hairyhum) [reply .github/workflows/main.yaml:33]: We do have the `checkout` step, which updates the code. `go.` files were synced before, but I'm not sure that's strictly required TBH. We could make it a check (that all deps are up to date) and don't move the `go.` files between jobs, which would make sense, but I just didn't want to introduce new behaviours in this PR. 
Doing `go mod download` and then packing and unpacking the files is way longer than to let each step (lint, build) to download deps themselves. Also if you run `go mod tidy` it does download all dependencies and there is no need to run `go mod download` again.

Comment (denisvmedia) [reply .github/workflows/main.yaml:33]: Let me join the questions left by @viveksinghggits . I'd also put some textual notes to the workflow to explain why we are doing that / what we are trying to achieve.

File: build/test.sh [hunk 2]
```diff
@@ -78,7 +78,6 @@ check_dependencies() {
 check_dependencies
 
 echo "Running tests:"
-go test -v -installsuffix "static" ${TARGETS}
```
Comment (denisvmedia) [inline build/test.sh:81]: Why is this removed?

Comment (hairyhum) [reply build/test.sh:81]: Originally we were builing tests with `go test -i` and then running it (next few lines), but then the flag was deprecated and removed from this line, but that meant that we run the tests twice, which doesn't make much sense.

**Review by denisvmedia — APPROVED** [review]
LGTM

**Review by viveksinghggits — APPROVED** [review]
